File: coursework/coursework1/seven_segments.py
from math import *
import numpy as np

fancy=False

def seven_segment(pattern):
    global fancy
    left,lower,right = \
        ('▕','__','▏') if fancy else \
        ('|','__','|')
    def vert(d1,d2,d3):
        print(
        (left  if d1 else " ")+
        (lower if d3 else "  ")+
        (right if d2 else " "))
    bits = [(ai==1) for ai in pattern]
    print(" "+lower+" " if bits[0] else "   ")
    vert(*bits[1:4])
    vert(*bits[4:7])
    number=0
    for i in range(0,4):
        if bits[7+i]:
            number+=pow(2,i)
    print('%X'%int(number))

hexdigits = np.int8([\
    [1,1,1,0,1,1,1, 0,0,0,0],
    [0,0,1,0,0,1,0, 1,0,0,0],
    [1,0,1,1,1,0,1, 0,1,0,0],
    [1,0,1,1,0,1,1, 1,1,0,0],
    [0,1,1,1,0,1,0, 0,0,1,0],
    [1,1,0,1,0,1,1, 1,0,1,0],
    [1,1,0,1,1,1,1, 0,1,1,0],
    [1,0,1,0,0,1,0, 1,1,1,0],
    [1,1,1,1,1,1,1, 0,0,0,1],
    [1,1,1,1,0,1,1, 1,0,0,1],
    [1,1,1,1,1,1,0, 0,1,0,1],
    [0,1,0,1,1,1,1, 1,1,0,1],
    [1,1,0,0,1,0,1, 0,0,1,1],
    [0,0,1,1,1,1,1, 1,0,1,1],
    [1,1,0,1,1,0,1, 0,1,1,1],
    [1,1,0,1,1,0,0, 1,1,1,1]])*2-1

zero, one, two, three, four, five, six, seven, eight, nine, hexA, hexB, hexC, hexD, hexE, hexF = hexdigits

test1 = np.int8([1,-1,1,1,-1,1,1,-1,-1,-1,-1])
test2 = np.int8([1,1,1,1,1,1,1,-1,-1,-1,-1])

if __name__=="__main__":
    for d in hexdigits: seven_segment(d)
    seven_segment(three)
    seven_segment(six)
    seven_segment(one)
    print("test1")
    seven_segment(test1)
    print("test2")
    seven_segment(test2)

File: coursework/coursework1/seven_segments.jl
function seven_segment(pattern::Array{Int64})
    function to_bool(a::Int64)
        return a==1
    end    
    function vert(d1,d2,d3)
        println(
            (d1 ? "|" : " ")*
            (d3 ? "_" : " ")*
            (d2 ? "|" : " "))
    end
    pattern_b=map(to_bool,pattern)
    println(pattern_b[1] ? " _ " : "   ")
    vert(pattern_b[2],pattern_b[3],pattern_b[4])
    vert(pattern_b[5],pattern_b[6],pattern_b[7])
    number=0
    for i in 0:3
        if pattern_b[7+i+1]
            number+=2^i
        end
    end
    println(number)
end

hexdigits = Matrix{Int8}([
    0 0 1 0 0 1 0 1 0 0 0 
    1 0 1 1 1 0 1 0 1 0 0 
    1 0 1 1 0 1 1 1 1 0 0 
    0 1 1 1 0 1 0 0 0 1 0 
    1 1 0 1 0 1 1 1 0 1 0 
    1 1 0 1 1 1 1 0 1 1 0 
    1 0 1 0 0 1 0 1 1 1 0 
    1 1 1 1 1 1 1 0 0 0 1 
    1 1 1 1 0 1 1 1 0 0 1 
    1 1 1 1 1 1 0 0 1 0 1 
    0 1 0 1 1 1 1 1 1 0 1 
    1 1 0 0 1 0 1 0 0 1 1 
    0 0 1 1 1 1 1 1 0 1 1 
    1 1 0 1 1 0 1 0 1 1 1 
    1 1 0 1 1 0 0 1 1 1 1
    1 1 1 0 1 1 1 0 0 0 0 ]) .* 2 .- 1;
one , two, three, four  = [ hexdigits[i,:] for i in  1:4 ];
five, six, seven, eight = [ hexdigits[i,:] for i in  5:8 ];
nine, hexA, hexB, hexC  = [ hexdigits[i,:] for i in  9:12];
hexD, hexE, hexF, zero  = [ hexdigits[i,:] for i in 13:16];

test1 = Int64[1,-1,1,1,-1,1,1,-1,-1,-1,-1]
test2 = Int64[1,1,1,1,1,1,1,-1,-1,-1,-1]

if abspath(PROGRAM_FILE) == @__FILE__
    seven_segment(three)
    seven_segment(six)
    seven_segment(one)
    println("test1")
    seven_segment(test1)
    println("test2")
    seven_segment(test2)
end